JBTM-3422 Publish LRA 1.0 TCK results on narayana.io

xstefank opened the pull request:
https://issues.redhat.com/browse/JBTM-3422

xstefank committed:
JBTM-3422 Publish LRA 1.0 TCK results on narayana.io

xstefank commented:
/cc @ochaloup @mmusgrov

tomjenkinson commented:
TESTIT

tomjenkinson commented:
Hi @xstefank - do they have to be zips?

mmusgrov reviewed (commented):
How do we publicise the results, there isn't any way to navigate to them. How about adding a link after the `Issue Tracker` item in the file `_site/index.html`:

111a112,114
>                 <li>
>                   <a href="certifications/microprofile/lra/1.0/CCR-results/index.html">Certifications</a>
>                 </li>

I also checked the links and they are okay. However, the instructions for building the project failed building wildfly:

```
[ERROR] Failed to execute goal on project wildfly-testsuite-shared: Could not resolve dependencies for project org.wildfly:wildfly-testsuite-shared:jar:24.0.0.Beta1-SNAPSHOT: Failure to find net.minidev:json-smart:jar:2.3 in https://repository.jboss.org/nexus/content/groups/public/ was cached in the local repository, resolution will not be reattempted until the update interval of jboss-public-repository-group has elapsed or updates are forced -> [Help 1]
```

Since we don't need wildfly how about changing the instruction `WORKSPACE=$PWD PROFILE=LRA ./scripts/hudson/narayana.sh` to the following:

`./build.sh clean install -DskipTests -Pcommunity`

xstefank commented:
@tomjenkinson, there are ~30 files that I wanted to include so I choose zip. If exploded list of 30 files would be preferable I will change it.

@mmusgrov 
> How do we publicise the results, there isn't any way to navigate to them. How about adding a link after the Issue Tracker item in the file _site/index.html:

I did this on purpose as this CCR is not a certification request for the Narayana as the implementation but a certification request that allows LRA specification to be released. There is not at least for now any motion to certify any MP implementation against any particular MP specification only you can certify the whole MP platform (8 specs in the umbrella profile) which is outside of the scope of the LRA. So this page can basically be "discarded" after LRA 1.0 release (of course we should keep it for completeness if someone in the future will want to find how LRA 1.0 was certified but officially it will have no value after the release). So we can link it if you prefer but I don't see the value outside the scope of the CCR from which this page will be directly linked. Let me know please.

> I also checked the links and they are okay. However, the instructions for building the project failed building wildfly:

> [ERROR] Failed to execute goal on project wildfly-testsuite-shared: Could not resolve dependencies for project org.wildfly:wildfly-testsuite-shared:jar:24.0.0.Beta1-SNAPSHOT: Failure to find net.minidev:json-smart:jar:2.3 in https://repository.jboss.org/nexus/content/groups/public/ was cached in the local repository, resolution will not be reattempted until the update interval of jboss-public-repository-group has elapsed or updates are forced -> [Help 1]

This is an error in your local maven environment.

> Since we don't need wildfly how about changing the instruction WORKSPACE=$PWD PROFILE=LRA ./scripts/hudson/narayana.sh to the following:

> ./build.sh clean install -DskipTests -Pcommunity

This is for sure a good concern but I would prefer not to change it for this particular CCR now as I did both runs on JDK 8 and 11 on my machine with clear environments using this command - ` WORKSPACE=$PWD PROFILE=LRA ./scripts/hudson/narayana.sh`. I used a custom local maven repository to really simulate a clean build so these two runs took the whole working day on my machine to download all dependencies and I would prefer not to repeat it for this CCR. But of course, I can but it would be an unnecessary delay.

Please let me know how to proceed so we can get the LRA 1.0 out :)

ochaloup reviewed (approved):
To me the current proposal seems good.

I like what @mmusgrov proposed that the results should be navigable from the https://narayana.io site - i.e. having the new link from the top level menu about LRA (adding a link after the `Issue Tracker item`). But such page should contain all information about Narayana LRA - not only results of the LRA 1.0 certification.
That way I think we don't not need to propose such change in this PR and we can create such page in in some next PR (it can be created in this PR as well, just it's the reason why I'm approving the PR).
I think we should have a place where we can put resources about LRA - I think the information about specification, probably summarized where to find documentation, blogposts, quickstarts. I would like to place the OpenAPI spec for Narayana coordinator somewhere. And other resource may come later.

mmusgrov commented:
> @tomjenkinson, there are ~30 files that I wanted to include so I choose zip. If exploded list of 30 files would be preferable I will change it.
> 
> @mmusgrov
> 
> > How do we publicise the results, there isn't any way to navigate to them. How about adding a link after the Issue Tracker item in the file _site/index.html:
> 
> I did this on purpose as this CCR is not a certification request for the Narayana as the implementation but a certification request that allows LRA specification to be released. There is not at least for now any motion to certify any MP implementation against any particular MP specification only you can certify the whole MP platform (8 specs in the umbrella profile) which is outside of the scope of the LRA. So this page can basically be "discarded" after LRA 1.0 release (of course we should keep it for completeness if someone in the future will want to find how LRA 1.0 was certified but officially it will have no value after the release). So we can link it if you prefer but I don't see the value outside the scope of the CCR from which this page will be directly linked. Let me know please.

Strange, in that case simply emailing them the same information would have been sufficient. But if that is the guidance then fair enough. 

> 
> > I also checked the links and they are okay. However, the instructions for building the project failed building wildfly:
> 
> > [ERROR] Failed to execute goal on project wildfly-testsuite-shared: Could not resolve dependencies for project org.wildfly:wildfly-testsuite-shared:jar:24.0.0.Beta1-SNAPSHOT: Failure to find net.minidev:json-smart:jar:2.3 in https://repository.jboss.org/nexus/content/groups/public/ was cached in the local repository, resolution will not be reattempted until the update interval of jboss-public-repository-group has elapsed or updates are forced -> [Help 1]
> 
> This is an error in your local maven environment.

If I get the error then there is a reasonable chance that whoever is processing the CCR will hit the same issue and if they do then it will delay the process (especially considering that wildfly is not required). But if you want to stick with telling them to build with our Jenkins CI script then go for it.

> 
> > Since we don't need wildfly how about changing the instruction WORKSPACE=$PWD PROFILE=LRA ./scripts/hudson/narayana.sh to the following:
> 
> > ./build.sh clean install -DskipTests -Pcommunity
> 
> This is for sure a good concern but I would prefer not to change it for this particular CCR now as I did both runs on JDK 8 and 11 on my machine with clear environments using this command - ` WORKSPACE=$PWD PROFILE=LRA ./scripts/hudson/narayana.sh`. I used a custom local maven repository to really simulate a clean build so these two runs took the whole working day on my machine to download all dependencies and I would prefer not to repeat it for this CCR. But of course, I can but it would be an unnecessary delay.

Ah right, there was a lot of work to be done to reach this point. Fair enough, I agree let's not risk breaking it.

> 
> Please let me know how to proceed so we can get the LRA 1.0 out :)

I will approve the PR so that you can proceed. Thanks for bringing this phase of the process to fruition.

xstefank commented:
@ochaloup @mmusgrov 
I changed the path to accommodate lra as the context root so the CCR is now available at `/lra/certifications/microprofile-lra-1.0/CCR-results/`. I also added the link to the header as Mike suggested. It is called now "Certifications" and it points to /lra/certifications. I would expect that this would change to "LRA" once we have more content there. Please feel free to suggest changes or merge this PR.

mmusgrov commented on _layouts/project.html.haml:
```diff
@@ -103,6 +103,8 @@
                 %li
                   %a(href="https://issues.jboss.org/browse/JBTM") Issue Tracker
+                %li
+                  %a(href="/lra/certifications") Certifications
```
The link needs to be relative with a class. So comparing with the others (such as Community, for example):

```
        %li{:class => page.output_path.start_with?("/lra") ? "active": ""}
            %a{:href => relative("/lra/certifications/index.html")} Certifications
```

Similarly for all the other references to the Certifications menu item.

mmusgrov commented on lra/certifications/index.html.haml:
```diff
@@ -0,0 +1,14 @@
+
+
+%h2
+  %a(href="/lra/certifications/microprofile-lra-1.0/CCR-results") LRA 1.0 Compatible Certification Request (CCR)
```
The link needs to be relative to the directory and with an index.html, my guess (although I haven't tried it) would be something similar to:

        %li{:class => page.output_path.start_with?("/lra") ? "active": ""}
            %a{:href => relative("/lra/certifications/microprofile-lra-1.0/CCR-results/index.html")} LRA 1.0 Compatible Certification Request (CCR)

mmusgrov replied:
@xstefank your last push didn't add index.html to the url so that when you click on hyperlink it takes the user to a directory files list page not the actual index.html

xstefank commented:
@mmusgrov thanks, updated, can you check it please?

xstefank commented:
@mmusgrov I changed it also for the relative links to the test results zips in the CCR itself.

mmusgrov commented:
The PR is in a build queue, I will need to wait for that before rechecking that the zip links are correct.

mmusgrov commented on lra/certifications/microprofile-lra-1.0/CCR-results.html.haml:
```diff
@@ -0,0 +1,275 @@
+  %li
+    %p{:class => page.output_path.start_with?("/lra") ? "active": ""}
+      Results failsafe-reports zip -
+      %a{:href => relative("/lra/certifications/microprofile-lra-1.0/lra-1.0-jdk-11-ccr-results.zip")} lra-1.0-jdk-11-ccr-results.zip
```
The directory `CCR-results` is missing from this href.

mmusgrov commented on lra/certifications/index.html.haml:
```diff
@@ -0,0 +1,14 @@
+
+
+%h2{:class => page.output_path.start_with?("/lra") ? "active": ""}
+  %a{:href => relative("/lra/certifications/microprofile-lra-1.0/CCR-results/")} LRA 1.0 Compatible Certification Request (CCR)
```
The href needs index.html adding (otherwise the link takes the user to directory listing.

mmusgrov replied:
Also I think you have deleted the directory `CCR-results`

xstefank replied:
basically, this works if you name the file CCR-results.html.haml then you can reference is in this way but I agree I'll change it to index.html

mmusgrov replied:
thanks

mmusgrov commented on lra/certifications/microprofile-lra-1.0/CCR-results.html.haml:
```diff
@@ -0,0 +1,275 @@
+  %li
+    %p{:class => page.output_path.start_with?("/lra") ? "active": ""}
+      Results failsafe-reports zip -
+      %a{:href => relative("/lra/certifications/microprofile-lra-1.0/lra-1.0-jdk-11-ccr-results.zip")} lra-1.0-jdk-11-ccr-results.zip
```
I don't know haml but we have a similar link to files which appear to have a different syntax. Eg

```
                <li class="list-group-item">
                  <a href="..//docs/specs/server-integration-guide.pdf">
                    <i class="icon-external-link"></i>
                    Narayana Server Integration Guide, v4.14.0, 2010
                  </a>
                </li>
```

I was unable to check it on my laptop because installing awestruct on my laptop fails badly.

If you are unable to install awestruct as well then the quickest way to verify the syntax would be to either
a) merge this PR and then we check. If it's wrong then we just raise another PR.
b) ask Ondra if he would be able to check out your PR and build it locally.

WDYT?

There is also file called `./_config/links.yml` which contains file links. I don't know what it's for or whether you need it for these two zips.

xstefank replied:
yes, awestruck was paint to run but I was successful and I am running it locally and verifying everything that I push with awestruck -d. I'll check if this should be updated but the current version works for sure.

xstefank replied:
@mmusgrov I wasn't able to do the relative links. All links even in the relative() directive/method need to start with / so they need to be absolute. The linked example also resolves the paths dynamically from the root /. I don't think that this is such a big issue as this path shouldn't change nevertheless as it will be linked in the official CCR.

mmusgrov replied:
Okay I'll approve it and merge. Thanks for this.
